File: app/src/main/java/com/simplyillustrate/simplyillustrate/models/PracticeQuestion.java
package com.simplyillustrate.simplyillustrate.models;

import java.util.List;

import android.os.Parcel;
import android.os.Parcelable;
import android.os.Parcelable.Creator;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

public class PracticeQuestion implements Parcelable {

    @SerializedName("tags")
    @Expose
    private List<String> tags = null;
    @SerializedName("incorrectAnswers")
    @Expose
    private List<String> incorrectAnswers = null;
    @SerializedName("_id")
    @Expose
    private String id;
    @SerializedName("user")
    @Expose
    private String user;
    @SerializedName("type")
    @Expose
    private String type;
    @SerializedName("difficulty")
    @Expose
    private String difficulty;
    @SerializedName("question")
    @Expose
    private String question;
    @SerializedName("correctAnswer")
    @Expose
    private String correctAnswer;
    @SerializedName("created_at")
    @Expose
    private String createdAt;
    @SerializedName("__v")
    @Expose
    private Integer v;
    public final static Parcelable.Creator<PracticeQuestion> CREATOR = new Creator<PracticeQuestion>() {


        @SuppressWarnings({
                "unchecked"
        })
        public PracticeQuestion createFromParcel(Parcel in) {
            return new PracticeQuestion(in);
        }

        public PracticeQuestion[] newArray(int size) {
            return (new PracticeQuestion[size]);
        }

    };

    protected PracticeQuestion(Parcel in) {
        in.readList(this.tags, (java.lang.String.class.getClassLoader()));
        in.readList(this.incorrectAnswers, (java.lang.String.class.getClassLoader()));
        this.id = ((String) in.readValue((String.class.getClassLoader())));
        this.user = ((String) in.readValue((String.class.getClassLoader())));
        this.type = ((String) in.readValue((String.class.getClassLoader())));
        this.difficulty = ((String) in.readValue((String.class.getClassLoader())));
        this.question = ((String) in.readValue((String.class.getClassLoader())));
        this.correctAnswer = ((String) in.readValue((String.class.getClassLoader())));
        this.createdAt = ((String) in.readValue((String.class.getClassLoader())));
        this.v = ((Integer) in.readValue((Integer.class.getClassLoader())));
    }

    public PracticeQuestion() {
    }

    public List<String> getTags() {
        return tags;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }

    public List<String> getIncorrectAnswers() {
        return incorrectAnswers;
    }

    public void setIncorrectAnswers(List<String> incorrectAnswers) {
        this.incorrectAnswers = incorrectAnswers;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getUser() {
        return user;
    }

    public void setUser(String user) {
        this.user = user;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getDifficulty() {
        return difficulty;
    }

    public void setDifficulty(String difficulty) {
        this.difficulty = difficulty;
    }

    public String getQuestion() {
        return question;
    }

    public void setQuestion(String question) {
        this.question = question;
    }

    public String getCorrectAnswer() {
        return correctAnswer;
    }

    public void setCorrectAnswer(String correctAnswer) {
        this.correctAnswer = correctAnswer;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }

    public Integer getV() {
        return v;
    }

    public void setV(Integer v) {
        this.v = v;
    }

    @Override
    public String toString() {
        return new StringBuilder().append("question = " + question)
                .append("correctAnswer = " + correctAnswer)
                .append("createdAt = " + createdAt).toString();
    }

    public void writeToParcel(Parcel dest, int flags) {
        dest.writeList(tags);
        dest.writeList(incorrectAnswers);
        dest.writeValue(id);
        dest.writeValue(user);
        dest.writeValue(type);
        dest.writeValue(difficulty);
        dest.writeValue(question);
        dest.writeValue(correctAnswer);
        dest.writeValue(createdAt);
        dest.writeValue(v);
    }

    public int describeContents() {
        return 0;
    }

}

File: app/src/main/java/com/simplyillustrate/simplyillustrate/entity/PracticeQuestion.kt
package com.simplyillustrate.simplyillustrate.entity

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import java.util.*

class PracticeQuestion() : Parcelable {

    @SerializedName("tags")
    var tags: List<Tag>? = null

    @SerializedName("incorrect_answers")
    var incorrectAnswers: List<String>? = null

    @SerializedName("_id")
    var id: String? = null

    @SerializedName("created_by")
    var createdBy: User? = null

    @SerializedName("type")
    var type: String? = null

    @SerializedName("difficulty")
    var difficulty: String? = null

    @SerializedName("question")
    var question: String? = null

    @SerializedName("correct_answer")
    var correctAnswer: String? = null

    @SerializedName("timestamp")
    var timestamp: Date? = null

    @SerializedName("__v")
    var v: Int? = null

    constructor(parcel: Parcel) : this() {
        tags = parcel.createTypedArrayList(Tag)
        incorrectAnswers = parcel.createStringArrayList()
        id = parcel.readString()
        createdBy = parcel.readParcelable(User::class.java.classLoader)
        type = parcel.readString()
        difficulty = parcel.readString()
        question = parcel.readString()
        correctAnswer = parcel.readString()
        timestamp = parcel.readSerializable() as? Date
        v = parcel.readValue(Int::class.java.classLoader) as? Int
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeTypedList(tags)
        parcel.writeStringList(incorrectAnswers)
        parcel.writeString(id)
        parcel.writeParcelable(createdBy, flags)
        parcel.writeString(type)
        parcel.writeString(difficulty)
        parcel.writeString(question)
        parcel.writeString(correctAnswer)
        parcel.writeSerializable(timestamp)
        parcel.writeValue(v)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<PracticeQuestion> {
        override fun createFromParcel(parcel: Parcel): PracticeQuestion {
            return PracticeQuestion(parcel)
        }

        override fun newArray(size: Int): Array<PracticeQuestion?> {
            return arrayOfNulls(size)
        }
    }
}
